src/avro_util.erl

```diff
@@ -295,6 +296,135 @@ resolve_duplicated_refs(Type0) ->
+  avro:get_type_fullname(AvroType).
+
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_RECORD_TYPE(Reader) andalso ?IS_RECORD_TYPE(Writer) ->
+  SameType = avro:is_same_type(Reader, Writer),
+  ReaderTypes = avro_record:get_all_field_data(Reader),
+  WriterTypes = avro_record:get_all_field_data(Writer),
+  NewRPath = [desc(Reader) | RPath],
+  NewWPath = [desc(Writer) | WPath],
+  SameType andalso
+    lists:all(
+      fun({FieldName, FieldType, Default}) ->
+          case lists:keysearch(FieldName, 1, WriterTypes) of
+            false ->
+              Default =/= ?NO_VALUE orelse
+                erlang:throw({ reader_missing_defalut_value
+                             , [{field, FieldName} |  NewWPath]
+                             });
+            {value, {_, WriterType, _}} ->
+              FieldDesc = {field, FieldName},
+              do_is_compatible(FieldType, WriterType,
+                               [FieldDesc | NewRPath],
+                               [FieldDesc | NewWPath])
+          end
+      end,
+      ReaderTypes);
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_ENUM_TYPE(Reader) andalso ?IS_ENUM_TYPE(Writer) ->
+  avro:is_same_type(Reader, Writer) andalso
+    Writer#avro_enum_type.symbols -- Reader#avro_enum_type.symbols == [] orelse
+    erlang:throw({ not_compatible, [desc(Reader) | RPath]
+                 , [desc(Writer) | WPath]});
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_FIXED_TYPE(Reader) andalso ?IS_FIXED_TYPE(Writer) ->
+  avro:is_same_type(Reader, Writer)
+    andalso (Reader#avro_fixed_type.size == Writer#avro_fixed_type.size) orelse
+    erlang:throw( {not_compatible, [desc(Reader) | RPath]
+                  , [desc(Writer) | WPath]});
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_UNION_TYPE(Reader) andalso ?IS_UNION_TYPE(Writer) ->
+  ReaderTypes = avro_union:get_types(Reader),
+  WriterTypes = avro_union:get_types(Writer),
+  NewRPath = [desc(Reader) | RPath],
+  NewWPath = [desc(Writer) | WPath],
+  CheckFun =
+    fun F([], _, _) ->
+        true;
+        F([WriterType | WT], [ReaderType | RT], Index) ->
+        MemberIndex = {member_id, Index},
+        do_is_compatible(ReaderType, WriterType,
+                         [MemberIndex | NewRPath],
+                         [MemberIndex | NewWPath])
+          andalso F(WT, RT, Index + 1)
+    end,
+  CheckFun(WriterTypes, ReaderTypes, 0);
+do_is_compatible(Reader, Writer, RPath, WPath)
+  when ?IS_UNION_TYPE(Reader) ->
+  ReaderTypes = avro_union:get_types(Reader),
+  ReaderTypeCount = length(ReaderTypes),
+  ReaderTypeIndices = lists:zip(lists:seq(0, ReaderTypeCount - 1), ReaderTypes),
+  NewRPath = [desc(Reader) | RPath],
+  NewWPath = [desc(Writer) | WPath],
+  lists:any(
+    fun({Index, ReaderType}) ->
+        try
+          do_is_compatible(ReaderType, Writer, [{member_id, Index} |NewRPath],
+                           WPath)
+        catch throw:{not_compatible, _, _} ->
```

Comment:
nit: `throw : _ ->` to catch both

Reply:
This makes it more explicit avoiding catching something other than expected. If you see no risk for it whatsover, I can do the throw:_.

Reply:
if you have test cases to cover both, i'm good. :)

Reply:
I rely on coverage to not lie :)